src/diamonds/facets/SnapshotsLibrary.sol

```diff
@@ -102,60 +104,102 @@ library SnapshotsLibrary {
+                    "snapshot heights too close in MadNet"
+                );
+            }
         }
 
+        // Lock rewards
         ParticipantsLibrary.ParticipantsStorage storage ps = ParticipantsLibrary.participantsStorage();
-        StakingLibrary.StakingStorage storage stakingS = StakingLibrary.stakingStorage();
-
-        for (uint idx=0; idx<ps.validators.length; idx++) {
-            if (msg.sender==ps.validators[idx]) {
-                StakingLibrary.lockRewardFor(ps.validators[idx], stakingS.rewardAmount + stakingS.rewardBonus, cs.epoch+2);
-            } else {
-                StakingLibrary.lockRewardFor(ps.validators[idx], stakingS.rewardAmount, cs.epoch+2);
+        {
+            StakingLibrary.StakingStorage storage stakingS = StakingLibrary.stakingStorage();
+            for (uint idx=0; idx<ps.validators.length; idx++) {
+                if (msg.sender==ps.validators[idx]) {
+                    StakingLibrary.lockRewardFor(ps.validators[idx], stakingS.rewardAmount + stakingS.rewardBonus, cs.epoch+2);
+                } else {
+                    StakingLibrary.lockRewardFor(ps.validators[idx], stakingS.rewardAmount, cs.epoch+2);
+                }
             }
         }
 
+        // Check if sender is the elected validator allowed to make the snapshot
+        require(
+            mayValidatorSnapshot(
+                int(ps.ethdkgDiamond.numberOfRegistrations()),
+                int(ps.ethdkgDiamond.validatorIndex(msg.sender)),
+                int(ethBlocksSinceLastSnapshot) - int(ss.snapshotDesperationDelay),
+                keccak256(_bclaims),
+                int(ss.snapshotDesperationFactor)
+            ),
+            "validator not among allowed"
+        );
+
+        // determine whether validators changed and thus ethDKG needs to be rerun
         bool reinitEthdkg;
         if (ss.validatorsChanged) {
             reinitEthdkg = true;
         }
         ss.validatorsChanged = false;
 
-        emit SnapshotTaken(chainId, cs.epoch, height, msg.sender, ss.validatorsChanged);
+        emit SnapshotTaken(chainId, cs.epoch, madHeight, msg.sender, ss.validatorsChanged);
 
         cs.epoch++;
 
         return reinitEthdkg;
     }
 
+    function mayValidatorSnapshot(int numValidators, int myIdx, int blocksSinceDesperation, bytes32 blsig, int desperationFactor) internal pure returns (bool) {        
+        int numValidatorsAllowed = 1;
+        
+        for (int i = blocksSinceDesperation; i >= 0;) {
+            i -= desperationFactor/numValidatorsAllowed;
+            numValidatorsAllowed++;
+            
+            if (numValidatorsAllowed > numValidators/3) break;
+        }
+        
+        uint rand = uint(blsig);
+        int start = int(rand % uint(numValidators));
+        int end = (start + numValidatorsAllowed) % numValidators;
+        
+        if (end > start) {
+            return myIdx >= start && myIdx < end;
```

Comment:
So, indexes on the ETHDKG start at 1 and go all the way to MAX_PARTICIPANT. Does this logic take this into account? Just wondering about the signals `>=` and `<`

Reply:
I'm talking about the new logic, I'm just wondering if we have to sub or sum1 somewhere to solve to account for this detail.